staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go

```diff
@@ -349,8 +373,87 @@ func (o *DebugOptions) generateDebugContainer(pod *corev1.Pod) *corev1.Ephemeral
 	return ec
 }
 
-// waitForEphemeralContainer watches the given pod until the ephemeralContainer is running
-func waitForEphemeralContainer(ctx context.Context, podClient corev1client.PodsGetter, ns, podName, ephemeralContainerName string) (*corev1.Pod, error) {
+// generatePodCopy takes a Pod and returns a copy and the debug container name of that copy
+func (o *DebugOptions) generatePodCopyWithDebugContainer(pod *corev1.Pod) (*corev1.Pod, string) {
+	copied := &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        o.CopyTo,
+			Namespace:   pod.Namespace,
+			Annotations: pod.Annotations,
+		},
+		Spec: *pod.Spec.DeepCopy(),
+	}
+	// change ShareProcessNamespace configuration only when commanded explicitly
+	if o.shareProcessedChanged {
+		copied.Spec.ShareProcessNamespace = pointer.BoolPtr(o.ShareProcesses)
+	}
+	if o.CopyLabels {
+		copied.Labels = pod.Labels
+	}
+	if !o.SameNode {
+		copied.Spec.NodeName = ""
+	}
+
+	containerByName := containerNameToRef(copied)
+	c, containerExists := containerByName[o.Container]
+	// Add a new container if the specified container does not exist
+	if !containerExists {
+		name := o.computeDebugContainerName(copied)
+		c = &corev1.Container{Name: name}
+		// envs are customizable when adding new container
+		c.Env = o.Env
+	}
+	c.Image = o.Image
+	c.ImagePullPolicy = o.PullPolicy
+	c.Stdin = o.Interactive
+	c.TerminationMessagePolicy = corev1.TerminationMessageReadFile
+	c.TTY = o.TTY
+	if o.ArgsOnly {
+		c.Args = o.Args
+	} else {
+		c.Command = o.Args
+		c.Args = nil
+	}
+	if !containerExists {
+		copied.Spec.Containers = append(copied.Spec.Containers, *c)
+	}
+	return copied, c.Name
+}
+
+func (o *DebugOptions) computeDebugContainerName(pod *corev1.Pod) string {
+	name := o.Container
+	if len(name) == 0 {
+		cn, containerByName := "", containerNameToRef(pod)
+		for len(cn) == 0 || (containerByName[cn] != nil) {
+			cn = fmt.Sprintf("debugger-%s", nameSuffixFunc(5))
+		}
+		if !o.Quiet {
+			fmt.Fprintf(o.ErrOut, "Defaulting debug container name to %s.\n", cn)
```

Comment:
I didn't catch it last time, but seems weird that it's on `ErrOut`.

Reply:
I think the correct behavior here is for output from `kubectl` and the container to use separate channels. For example, `kubectl debug mypod --image=busybox -- ifconfig -a > interfaces.txt` shouldn't write "Defaulting debug container..." to interfaces.txt.

Reply:
That's where you could use `-v=1` instead, or `--quiet` rather than printing on stderr something that is not an error.

Reply:
It's worth printing the message by default to make it easier to reconnect (and in the case of generating a pod name for node debugging, there's no other way to tell what the name is), but it's a good point that you can accomplish the same thing with `--quiet`.

The Unix user in me still thinks that stderr is correct here, but it's more important that `kubectl` is consistent. I can update this in my pending PR.